File: packages/ts/src/authz/primitives.ts
/**
 * Typed permission primitives — Permission, Role, Relation, RelationTuple.
 *
 * Supports boolean composition via .and() / .or():
 *   memberOf.and(editor).or(admin)
 *   canRead.and(canWrite)
 *
 * Permission format is separator-agnostic — "user:read" and "user.read"
 * are semantically equal. Supported separators: @ # . : | \ / $ &
 */

import type { AuthContext } from "../context";
import type { Action, Resource } from "../actions";
import { Requirement } from "./requirement";

export type { Action, Resource } from "../actions";

// ── Separator auto-detection ──────────────────────────────────────

const SEP_RE = /[@#.:|\\/$&]/;

// ── Shared permission matching ────────────────────────────────────

/**
 * Check if a permission pattern matches a target permission.
 *
 * Performs semantic (resource, action) comparison — separator-agnostic.
 *
 * Supports:
 * - Exact match: `"user:read"` matches `"user.read"`
 * - Global wildcard: `"*"` matches everything
 * - Resource wildcard: `"user:*"` matches `"user:read"`, `"user.write"`, etc.
 */
export function matchPermission(
  pattern: Permission | string,
  target: Permission | string,
): boolean {
  const p = typeof pattern === "string" ? new Permission(pattern) : pattern;
  const t = typeof target === "string" ? new Permission(target) : target;
  if (String(p.resource) === "*") return true;
  if (String(p.resource) !== String(t.resource)) return false;
  return String(p.action) === "*" || String(p.action) === String(t.action);
}

// ── Primitives ──────────────────────────────────────────────────

export type PermissionParser = (s: string) => [string, string];

/**
 * A typed permission combining a resource and an action.
 *
 * Accepts two args, a single string (auto-detects separator), or a custom parser:
 *
 *     new Permission("user", "read")      // two-arg form
 *     new Permission("user:read")          // colon separator (auto-detected)
 *     new Permission("user.read")          // dot separator (auto-detected)
 *     new Permission("*")                  // global wildcard
 *
 * Equality is semantic: `Permission("user:read").equals("user.read")` is true.
 * Supports `&` / `|` composition with other requirements via `.and()` / `.or()`.
 */
export class Permission extends Requirement {
  readonly resource: Resource;
  readonly action: Action;
  private _sep: string;

  constructor(
    resource: Resource | string,
    action?: Action | string,
    options?: { separator?: string; parser?: PermissionParser },
  ) {
    super();
    if (action !== undefined) {
      this.resource = String(resource) as Resource;
      this.action = String(action) as Action;
      this._sep = options?.separator ?? ":";
    } else if (options?.parser) {
      const [r, a] = options.parser(String(resource));
      this.resource = r as Resource;
      this.action = a as Action;
      this._sep = options.separator ?? ":";
    } else {
      const value = String(resource);
      if (value === "*") {
        this.resource = "*" as Resource;
        this.action = "*" as Action;
        this._sep = options?.separator ?? ":";
      } else {
        const m = SEP_RE.exec(value);
        if (!m) {
          throw new Error(
            `No separator found in permission string: "${value}". ` +
              `Use one of: @ # . : | \\ / $ &`,
          );
        }
        const parts = value.split(m[0]!);
        this.resource = parts[0]! as Resource;
        this.action = parts.slice(1).join(m[0]!) as Action;
        this._sep = m[0]!;
      }
    }
  }

  evaluate(ctx: AuthContext): boolean {
    return ctx.hasPermission(this);
  }

  toString(): string {
    return `${this.resource}${this._sep}${this.action}`;
  }

  equals(other: Permission | string): boolean {
    if (typeof other === "string") {
      try {
        const p = new Permission(other);
        return (
          String(this.resource) === String(p.resource) &&
          String(this.action) === String(p.action)
        );
      } catch {
        return false;
      }
    }
    return (
      String(this.resource) === String(other.resource) &&
      String(this.action) === String(other.action)
    );
  }
}

/**
 * Static role definition with associated permissions.
 *
 * Compares equal to its name string: `new Role("admin").equals("admin")`.
 */
export class Role extends Requirement {
  readonly name: string;
  readonly permissions: Permission[];

  constructor(name: string, permissions?: Permission[]) {
    super();
    this.name = name;
    this.permissions = permissions ?? [];
  }

  evaluate(ctx: AuthContext): boolean {
    return ctx.hasRole(this);
  }

  toString(): string {
    return this.name;
  }

  equals(other: Role | string): boolean {
    const otherName = typeof other === "string" ? other : other.name;
    return this.name === otherName;
  }
}

export type RelationParser = (s: string) => [string, string];

/**
 * Zanzibar-style relation definition.
 *
 * Resource-first ordering matches Python and the string form:
 *
 *     new Relation("doc", "owner")     // two-arg form (resource, name)
 *     new Relation("doc#owner")        // string form (auto-detects separator)
 *
 * Equality is semantic: `Relation("doc#owner").equals("doc.owner")` is true.
 * Supports `.and()` / `.or()` composition with other requirements.
 */
export class Relation extends Requirement {
  readonly resource: Resource;
  readonly name: string;
  private _sep: string;

  constructor(
    resource: Resource | string,
    name?: string,
    options?: { separator?: string; parser?: RelationParser },
  ) {
    super();
    if (name !== undefined) {
      this.resource = String(resource) as Resource;
      this.name = String(name);
      this._sep = options?.separator ?? "#";
    } else if (options?.parser) {
      const [r, n] = options.parser(String(resource));
      this.resource = r as Resource;
      this.name = n;
      this._sep = options.separator ?? "#";
    } else {
      const value = String(resource);
      const m = SEP_RE.exec(value);
      if (!m) {
        throw new Error(
          `No separator found in relation string: "${value}". ` +
            `Use one of: @ # . : | \\ / $ &`,
        );
      }
      const parts = value.split(m[0]!);
      this.resource = parts[0]! as Resource;
      this.name = parts.slice(1).join(m[0]!);
      this._sep = m[0]!;
    }
  }

  get separator(): string {
    return this._sep;
  }

  /** Create a full Zanzibar tuple from this relation definition. */
  tuple(objectId: string, subject?: string): RelationTuple {
    return new RelationTuple(this, objectId, subject);
  }

  evaluate(ctx: AuthContext): boolean {
    return ctx.relations.some(
      (rt) => rt.relation.name === this.name && String(rt.relation.resource) === String(this.resource),
    );
  }

  toString(): string {
    return `${this.resource}${this._sep}${this.name}`;
  }

  equals(other: Relation | string): boolean {
    if (typeof other === "string") {
      try {
        const r = new Relation(other);
        return this.name === r.name && String(this.resource) === String(r.resource);
      } catch {
        return false;
      }
    }
    return this.name === other.name && String(this.resource) === String(other.resource);
  }
}

/**
 * Full Zanzibar relationship tuple: `object_type:object_id#relation@subject`.
 *
 *     new RelationTuple(new Relation("doc", "owner"), "readme", "user:alice")
 *     RelationTuple.parse("doc:readme#owner@user:alice")
 */
export class RelationTuple {
  readonly relation: Relation;
  readonly objectId: string;
  readonly subject: string | undefined;

  constructor(relation: Relation, objectId: string, subject?: string) {
    this.relation = relation;
    this.objectId = objectId;
    this.subject = subject;
  }

  /** Parse `'doc:readme#owner@user:alice'` positionally by separator and `@`. */
  static parse(s: string): RelationTuple {
    let left: string;
    let subject: string | undefined;
    if (s.includes("@")) {
      const atIdx = s.indexOf("@");
      left = s.slice(0, atIdx);
      subject = s.slice(atIdx + 1);
    } else {
      left = s;
      subject = undefined;
    }
    // left = "doc:readme#owner"
    const colonPos = left.indexOf(":");
    if (colonPos === -1) {
      throw new Error(`Invalid relation tuple: "${s}"`);
    }
    const objType = left.slice(0, colonPos);
    const rest = left.slice(colonPos + 1); // "readme#owner"
    const m = SEP_RE.exec(rest);
    if (!m) {
      throw new Error(`Invalid relation tuple: "${s}"`);
    }
    const objId = rest.slice(0, m.index);
    const relName = rest.slice(m.index! + m[0]!.length);
    return new RelationTuple(
      new Relation(objType, relName),
      objId,
      subject,
    );
  }

  toString(): string {
    const sep = this.relation.separator;
    const base = `${this.relation.resource}:${this.objectId}${sep}${this.relation.name}`;
    if (this.subject !== undefined) {
      return `${base}@${this.subject}`;
    }
    return base;
  }

  equals(other: RelationTuple | string): boolean {
    if (typeof other === "string") {
      return this.toString() === other;
    }
    return (
      this.relation.equals(other.relation) &&
      this.objectId === other.objectId &&
      this.subject === other.subject
    );
  }
}

File: packages/rust/src/authz/primitives.rs
use std::fmt;

const SEPARATORS: &[char] = &[':', '.', '@', '#', '|', '/', '$', '&'];

// ---------------------------------------------------------------------------
// Permission
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, PartialEq)]
pub struct Permission {
    pub resource: String,
    pub action: String,
}

impl Permission {
    pub fn new(resource: &str, action: &str) -> Self {
        Self {
            resource: resource.to_string(),
            action: action.to_string(),
        }
    }

    /// Parse a permission string using any of the supported separators.
    ///
    /// Accepted forms:
    ///   "*"             -> resource="*", action="*"
    ///   "resource:*"    -> resource="resource", action="*"
    ///   "resource:read" -> resource="resource", action="read"
    ///   "resource"      -> resource="resource", action="*"
    pub fn parse(s: &str) -> Self {
        let s = s.trim();
        if s == "*" {
            return Self::new("*", "*");
        }
        for &sep in SEPARATORS {
            if let Some(pos) = s.find(sep) {
                let resource = &s[..pos];
                let action = &s[pos + sep.len_utf8()..];
                return Self::new(resource, action);
            }
        }
        // No separator found – treat the whole string as a resource with wildcard action.
        Self::new(s, "*")
    }

    /// Semantic matching with wildcard support.
    ///
    /// `self` is the *pattern*, `other` is the *target* being checked.
    ///   - `*:*`  matches everything
    ///   - `docs:*` matches any action on the `docs` resource
    ///   - Exact match is case-sensitive
    pub fn matches(&self, other: &Permission) -> bool {
        let resource_ok = self.resource == "*" || self.resource == other.resource;
        let action_ok = self.action == "*" || self.action == other.action;
        resource_ok && action_ok
    }
}

impl From<&str> for Permission {
    fn from(s: &str) -> Self {
        Permission::parse(s)
    }
}

impl fmt::Display for Permission {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}:{}", self.resource, self.action)
    }
}

/// Convenience: check whether a pattern permission string matches a target.
pub fn match_permission(pattern: &str, target: &str) -> bool {
    let pat = Permission::parse(pattern);
    let tgt = Permission::parse(target);
    pat.matches(&tgt)
}

// ---------------------------------------------------------------------------
// Role
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, PartialEq)]
pub struct Role {
    pub name: String,
    pub permissions: Vec<Permission>,
}

impl Role {
    pub fn new(name: &str, permissions: Vec<Permission>) -> Self {
        Self {
            name: name.to_string(),
            permissions,
        }
    }
}

impl fmt::Display for Role {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.name)
    }
}

// ---------------------------------------------------------------------------
// Relation
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, PartialEq)]
pub struct Relation {
    pub resource: String,
    pub name: String,
}

impl Relation {
    pub fn new(resource: &str, name: &str) -> Self {
        Self {
            resource: resource.to_string(),
            name: name.to_string(),
        }
    }

    /// Parse `"resource#name"` or `"resource:name"`.
    pub fn parse(s: &str) -> Self {
        let s = s.trim();
        if let Some(pos) = s.find('#') {
            let resource = &s[..pos];
            let name = &s[pos + 1..];
            return Self::new(resource, name);
        }
        if let Some(pos) = s.find(':') {
            let resource = &s[..pos];
            let name = &s[pos + 1..];
            return Self::new(resource, name);
        }
        Self::new(s, "")
    }

    /// Create a full relation tuple with an object id and optional subject.
    pub fn tuple(&self, object_id: &str, subject: Option<&str>) -> RelationTuple {
        RelationTuple {
            relation: self.clone(),
            object_id: object_id.to_string(),
            subject: subject.map(|s| s.to_string()),
        }
    }
}

impl From<&str> for Relation {
    fn from(s: &str) -> Self {
        Relation::parse(s)
    }
}

impl fmt::Display for Relation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}#{}", self.resource, self.name)
    }
}

// ---------------------------------------------------------------------------
// RelationTuple
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, PartialEq)]
pub struct RelationTuple {
    pub relation: Relation,
    pub object_id: String,
    pub subject: Option<String>,
}

impl RelationTuple {
    /// Parse the Zanzibar-style format `"resource:object_id#relation@subject"`.
    ///
    /// Examples:
    ///   `"doc:readme#owner@user:alice"`
    ///   `"folder:root#viewer"`  (no subject)
    pub fn parse(s: &str) -> Option<Self> {
        let s = s.trim();

        // Split resource:object_id from relation(@subject)?
        let colon_pos = s.find(':')?;
        let resource = &s[..colon_pos];
        let rest = &s[colon_pos + 1..];

        // rest = "readme#owner@user:alice" or "readme#owner"
        let hash_pos = rest.find('#')?;
        let object_id = &rest[..hash_pos];
        let after_hash = &rest[hash_pos + 1..];

        let (relation_name, subject) = if let Some(at_pos) = after_hash.find('@') {
            let rel = &after_hash[..at_pos];
            let subj = &after_hash[at_pos + 1..];
            (rel, Some(subj.to_string()))
        } else {
            (after_hash, None)
        };

        Some(Self {
            relation: Relation::new(resource, relation_name),
            object_id: object_id.to_string(),
            subject,
        })
    }
}

impl fmt::Display for RelationTuple {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}:{}#{}", self.relation.resource, self.object_id, self.relation.name)?;
        if let Some(ref subject) = self.subject {
            write!(f, "@{}", subject)?;
        }
        Ok(())
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn permission_wildcard_matches_everything() {
        let pat = Permission::new("*", "*");
        let target = Permission::new("docs", "read");
        assert!(pat.matches(&target));
    }

    #[test]
    fn permission_resource_wildcard() {
        let pat = Permission::new("docs", "*");
        assert!(pat.matches(&Permission::new("docs", "read")));
        assert!(pat.matches(&Permission::new("docs", "write")));
        assert!(!pat.matches(&Permission::new("users", "read")));
    }

    #[test]
    fn permission_exact_match() {
        let pat = Permission::new("docs", "read");
        assert!(pat.matches(&Permission::new("docs", "read")));
        assert!(!pat.matches(&Permission::new("docs", "write")));
    }

    #[test]
    fn permission_parse_various_separators() {
        assert_eq!(Permission::parse("docs:read"), Permission::new("docs", "read"));
        assert_eq!(Permission::parse("docs.read"), Permission::new("docs", "read"));
        assert_eq!(Permission::parse("docs@read"), Permission::new("docs", "read"));
        assert_eq!(Permission::parse("docs#read"), Permission::new("docs", "read"));
        assert_eq!(Permission::parse("*"), Permission::new("*", "*"));
        assert_eq!(Permission::parse("docs"), Permission::new("docs", "*"));
    }

    #[test]
    fn match_permission_convenience() {
        assert!(match_permission("docs:*", "docs:read"));
        assert!(!match_permission("docs:write", "docs:read"));
    }

    #[test]
    fn relation_parse() {
        let rel = Relation::parse("doc#owner");
        assert_eq!(rel.resource, "doc");
        assert_eq!(rel.name, "owner");
    }

    #[test]
    fn relation_tuple_parse_with_subject() {
        let tuple = RelationTuple::parse("doc:readme#owner@user:alice").unwrap();
        assert_eq!(tuple.relation.resource, "doc");
        assert_eq!(tuple.relation.name, "owner");
        assert_eq!(tuple.object_id, "readme");
        assert_eq!(tuple.subject, Some("user:alice".to_string()));
    }

    #[test]
    fn relation_tuple_parse_without_subject() {
        let tuple = RelationTuple::parse("folder:root#viewer").unwrap();
        assert_eq!(tuple.relation.resource, "folder");
        assert_eq!(tuple.relation.name, "viewer");
        assert_eq!(tuple.object_id, "root");
        assert_eq!(tuple.subject, None);
    }

    #[test]
    fn relation_tuple_roundtrip() {
        let input = "doc:readme#owner@user:alice";
        let tuple = RelationTuple::parse(input).unwrap();
        assert_eq!(tuple.to_string(), input);
    }
}
